# METI

## METI: Deep profiling of tumor ecosystems by integrating cell morphology and spatial transcriptomics

### Jiahui Jiang*, Yunhe Liu, Jiangjiang Qin, Jingjing Wu, Jianfeng Chen, Melissa P. Pizzi, Rossana L. Segura, Kohei Yamashita, Zhiyuan Xu, Guangsheng Pei, Kyung Serk Cho, Yanshuo Chu, Ansam F. Sinjab, Fuduan Peng, Guangchun Han, Ruiping Wang, Xinmiao Yan, Enyu Dai, Yibo Dai1, Mingyao Li, Andrew Futreal, Anirban Maitra, Alexander Lazar, Xiangdong Cheng, Humam Kadara, Jaffer Ajani, Amir A. Jazaeri, Jianjun Gao, Jian Hu*, Linghua Wang*

METI (Morphology-Enhanced Spatial Transcriptome Analysis Integrator) is an novel analytic framework that systematically analyzes cancer cells and cells of the TME by incorporating spatial gene expression, tissue histology, and prior knowledge of cancer and TME cells. METI starts with the identification of key cellular components and their states within the TME, including various immune cells and their transcriptional states, tumor stromal components such as cancer-associated fibroblasts (CAFs), and the epithelial compartment. Meanwhile, METI offers complementary information on cell morphology for various cell type from the H&E images. The combined results from gene expression and histology features provide a comprehensive understanding of the spatial cellular composition and organization within the tissue. The evaluation of METI shows robust and consistent performance across ST datasets generated from diverse cancer types including gastric cancer, lung cancer, and bladder cancer.

<br>

![METI workflow]()

<br>



